python/pyarrow/io.pxi

```diff
@@ -989,9 +1035,67 @@ cdef class BufferReader(NativeFile):
         else:
             self.buffer = py_buffer(obj)
 
-        self.rd_file.reset(new CBufferReader(self.buffer.buffer))
+        self.set_random_access_file(shared_ptr[RandomAccessFile](
+            new CBufferReader(self.buffer.buffer)))
         self.is_readable = True
-        self.closed = False
+
+
+cdef class CompressedInputStream(NativeFile):
+    """
+    An input stream wrapper which decompresses data on the fly.
+
+    Parameters
+    ----------
+    stream : pa.NativeFile
+    compression : str
+        The compression type (e.g. "gzip" or "bz2")
+    """
+    def __cinit__(self, NativeFile stream, compression):
```

Comment:
Using `__cinit__` is unsafe if passing no arguments is not an option (will segfault). Suggest we use a factory function to construct or similar. Speaking of which `pa.BufferReader` has the same issue

Reply:
This must be a bug that was fixed in a recent version of Cython, glad it's no longer an issue